File: scala/scalasearch/src/main/scala/scalasearch/SearchMain.scala
package scalasearch

import scalafind.{Common, FindException, Finder}

object SearchMain {
  def main(args: Array[String]): Unit = {
    var colorize = true
    var searchOptions: Option[SearchOptions] = None

    try {
      val config = new SearchConfig()
      searchOptions = Some(new SearchOptions(config))
      val settings = searchOptions.get.settingsFromArgs(args)
      colorize = settings.colorize

      if (settings.debug) {
        Common.log("settings: " + settings)
      }

      if (settings.printUsage) {
        Common.log("")
        searchOptions.foreach(_.usage(0))
      }

      val searcher = new Searcher(config, settings)
      val results = searcher.search()
      val formatter = new SearchResultFormatter(settings)

      if (settings.printResults) {
        Searcher.printSearchResults(results, formatter)
      }
      if (settings.printDirs || settings.printFiles) {
        val fileResults = Searcher.getMatchingFileResults(results)
        if (settings.printDirs) {
          Finder.printMatchingDirs(fileResults, formatter.fileResultFormatter)
        }
        if (settings.printFiles) {
          Finder.printMatchingFiles(fileResults, formatter.fileResultFormatter)
        }
      }
      if (settings.printLines) { searcher.printMatchingLines(results, formatter) }
      if (settings.printMatches) { searcher.printMatches(results, formatter) }

    } catch {
      case e: FindException =>
        Common.log("")
        Common.logError(e.getMessage + "\n", colorize)
        searchOptions.foreach(_.usage(1))
      case e: SearchException =>
        Common.log("")
        Common.logError(e.getMessage + "\n", colorize)
        searchOptions.foreach(_.usage(1))
    }
  }
}

File: fsharp/FsSearch/FsSearch/SearchMain.fs
namespace FsSearch

open FsFindLib
open FsSearchLib

module Main =

    let HandleError (err : string) (colorize : bool) (searchOptions : SearchOptions) : unit =
        Logger.Log("");
        Logger.LogErrorColor err colorize
        searchOptions.Usage(1)

    let Search (config : SearchConfig) (searchOptions : SearchOptions) (settings : SearchSettings) : unit =
        let searcher = Searcher(config, settings)

        let errs = searcher.ValidateSettings()
        if errs.Length > 0 then
            HandleError errs.Head settings.Colorize searchOptions

        let results = searcher.Search()
        let formatter = SearchResultFormatter(settings)

        if settings.PrintResults then
            searcher.PrintResults results formatter

        if settings.PrintDirs then
            searcher.PrintMatchingDirs results formatter

        if settings.PrintFiles then
            searcher.PrintMatchingFiles results formatter

        if settings.PrintLines then
            searcher.PrintMatchingLines results formatter

        if settings.PrintMatches then
            searcher.PrintMatches results formatter


    [<EntryPoint>]
    let Main(args : string[]) = 
        let config = SearchConfig()
        let searchOptions = SearchOptions(config)
        match searchOptions.SettingsFromArgs(args) with
        | Ok settings ->
            if settings.Debug then
                Logger.Log settings.ToString
            if settings.PrintUsage then
                searchOptions.Usage(0)
            else
                Search config searchOptions settings
        | Error e -> HandleError e true searchOptions

        // main entry point return
        0;;
